homeassistant/components/music_assistant/services.py

```diff
@@ -0,0 +1,91 @@
+"""Custom services for the Music Assistant integration."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, cast
+
+from music_assistant_client.helpers import searchresults_as_compact_dict
+from music_assistant_models.enums import MediaType
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+    callback,
+)
+from homeassistant.exceptions import HomeAssistantError
+import homeassistant.helpers.config_validation as cv
+
+from .const import DOMAIN
+
+if TYPE_CHECKING:
+    from music_assistant_client import MusicAssistantClient
+
+    from . import MusicAssistantConfigEntry
+
+SERVICE_SEARCH = "search"
+ATTR_MEDIA_TYPE = "media_type"
+ATTR_SEARCH_NAME = "name"
+ATTR_SEARCH_ARTIST = "artist"
+ATTR_SEARCH_ALBUM = "album"
+ATTR_LIMIT = "limit"
+ATTR_LIBRARY_ONLY = "library_only"
+
+
+@callback
+def get_music_assistant_client(hass: HomeAssistant) -> MusicAssistantClient:
+    """Get the (first) Music Assistant client from the (loaded) config entries."""
+    entry: MusicAssistantConfigEntry
+    for entry in hass.config_entries.async_entries(DOMAIN, False, False):
+        if entry.state != ConfigEntryState.LOADED:
+            continue
+        return entry.runtime_data.mass
+    raise HomeAssistantError("Music Assistant is not loaded")
+
+
+@callback
+def register_services(hass: HomeAssistant) -> None:
+    """Register custom services."""
+
+    async def handle_search(call: ServiceCall) -> ServiceResponse:
+        """Handle queue_command service."""
+        mass = get_music_assistant_client(hass)
+        search_name = call.data[ATTR_SEARCH_NAME]
+        search_artist = call.data.get(ATTR_SEARCH_ARTIST)
+        search_album = call.data.get(ATTR_SEARCH_ALBUM)
+        if search_album and search_artist:
+            search_name = f"{search_artist} - {search_album} - {search_name}"
+        elif search_album:
+            search_name = f"{search_album} - {search_name}"
+        elif search_artist:
+            search_name = f"{search_artist} - {search_name}"
+        search_results = await mass.music.search(
+            search_query=search_name,
+            media_types=call.data.get(ATTR_MEDIA_TYPE, MediaType.ALL),
+            limit=call.data[ATTR_LIMIT],
+            library_only=call.data[ATTR_LIBRARY_ONLY],
+        )
+        # return limited result to prevent it being too verbose
+        return cast(ServiceResponse, searchresults_as_compact_dict(search_results))
```

Comment:
Why do we cast? I don't recall that we have to cast

Reply:
mypy was being annoying to me without the cast. I can retry it now, but pretty sure it bails.

Reply:
Difference here is that the dict is created by mushumaro - I'll dig into it later today

Reply:
but that's why I use `asdict`

Reply:
asdict is not optimized, mushumaro handles that for you. You shouldnt use asdict on a dataclass handled by mashumaro.

Anyways, issue is that the typing in the mass client has dict[str, Any]:

![Scherm­afbeelding 2024-11-21 om 21 46 57](https://github.com/user-attachments/assets/d096627e-8c52-434d-af70-3ec2f9a71b03)


![Scherm­afbeelding 2024-11-21 om 21 47 44](https://github.com/user-attachments/assets/527646d8-fc72-41cb-8a21-b385d30b938b)

I'll fix that in a future update of the mass client